src/auth/policy/resolver.go

```diff
@@ -1,41 +1,349 @@
+	}
+
+	return nil
+}
+
+// actionsResolver is associated to every statement path
+// it aggregates all statement actions associated to the path into a radix tree
+type actionsResolver struct {
+	// exactActions maps an exact action name to an effect resolver
+	exactActions map[string]*effectResolver
+
+	// prefixedActions maps prefixed action to an effect resolver
+	prefixedActions *radix.Tree
+}
+
+func newActionsResolver() *actionsResolver {
+	return &actionsResolver{
+		exactActions:    make(map[string]*effectResolver),
+		prefixedActions: radix.New(),
+	}
+}
+
+func (rslvr *actionsResolver) addStatement(sttmnt *statement) error {
+	// loop over all actions in the statement
+	for _, action := range sttmnt.Actions {
+		if strings.HasSuffix(action, wildCardSuffix) {
+			// action is a prefix pattern (e.g. Sign*)
+			actionPrefix := strings.TrimSuffix(action, wildCardSuffix)
+
+			// lookup for effect already associtated to this prefix
+			effect, ok := rslvr.prefixedActions.Get(actionPrefix)
+			if ok {
+				// we already had an effect for prefix so we acculmulate the statement
+				err := effect.(*effectResolver).addStatement(sttmnt)
+				if err != nil {
+					return err
+				}
+			} else {
+				// we create the effect and insert it into the radix tree
+				newEffect := new(effectResolver)
+				err := newEffect.addStatement(sttmnt)
+				if err != nil {
+					return err
+				}
+				_, _ = rslvr.prefixedActions.Insert(actionPrefix, newEffect)
+			}
+		} else {
+			// action is an exact name (e.g. SignTransaction)
+
+			// lookup for effect already associated to this action
+			effect, ok := rslvr.exactActions[action]
+			if ok {
+				// we already had an effect for prefix so we acculmulate the statement
+				err := effect.addStatement(sttmnt)
+				if err != nil {
+					return err
+				}
+			} else {
+				// we create the effect and insert it into the map of exact actions
+				newEffect := new(effectResolver)
+				err := newEffect.addStatement(sttmnt)
+				if err != nil {
+					return err
+				}
+				rslvr.exactActions[action] = newEffect
+			}
+		}
+	}
+
+	return nil
+}
+
+func (rslvr *actionsResolver) isAuthorized(op *Operation) (res *result) {
+	res = &result{
+		op: op,
+	}
+
+	// do we have an exact match for the action?
+	actionEffects, hasExact := rslvr.exactActions[op.Action]
+	if hasExact {
+		// if we have an exact match and it resolves to Deny we return Deny
+		res = actionEffects.isAuthorized(op)
+		if res.deny != nil {
+			return res
+		}
+	}
+
+	// do we have a prefix matching the action?
+	actionPrefix, _, hasPrefix := rslvr.prefixedActions.LongestPrefix(op.Action)
+	if !hasPrefix {
+		// no prefix so we return
+		return res
+	}
+
+	// we have a prefix
+	// we walk all matching prefix starting from root up to the longest matching prefix
+	// if we meet a single Deny we stop walking and return a Deny
+	rslvr.prefixedActions.WalkPath(actionPrefix, func(_ string, v interface{}) bool {
```

Comment:
Reading Radix docs (see below) this method appear to navigate top to bottom which implies you will be apply policies `left-to-right`, but using the exactPath you are applying the oppositive behavior of reading `right-to-left` (see comment here https://github.com/ConsenSys/quorum-key-manager/pull/216/files#r671375931) , is there something I am missing here? 

```
// WalkPath is used to walk the tree, but only visiting nodes
// from the root down to a given leaf. Where WalkPrefix walks
// all the entries *under* the given prefix, this walks the
// entries *above* the given prefix.
func (t *Tree) WalkPath(path string, fn WalkFn) {
```